flips/20220715-cadence-purity-analysis.md

```diff
@@ -0,0 +1,211 @@
+need arise. 
+
+### Performance Implications
+
+This will add additional work to the type checking phase, so we will need to take care to ensure that this work is fast. 
+
+### Best Practices
+
+For the most part users can ignore this new feature unless they want to call functions from inside function conditions, in which case they will need to add
+`pure` annotations to those functions. However, if users have other reasons for wanting to limit what mutations can be performed from inside functions they
+may choose to add `pure` annotations anyways. 
+
+### User Impact
+
+This will have little direct impact on users, as `pure` annotations will only be required on functions that are called from conditions. This is, however, 
+technically backwards incompatible, since users will need to add annotations to these functions in order for their code to typecheck. 
+
+## Prior Art
+
+A number of other languages have similar features or related analyses from which we can learn. 
+
+C++ has both the `pure` and the `const` keyword. `pure` functions can read from, but not modify,
+any observable program state, while `const` functions can neither read nor modify any externally
+observable state. C++'s `const` is likely more powerful than anything we'd need in Cadence, as 
+capturing or reading global state does not have the same kind of safety concerns as modifying it. 
+Note that C++ does not do any kind of purity inference, hence why `const` has a poor reputation 
+among developers for requiring annotations all over the codebase. 
+
+D has `pure` functions that behave like `const` in C++, in that they cannot access or modify
+any kind of global or static mutable state. Like C++, D does not perform any purity analyses. 
+
+Solidity has `view` and `pure` functions, which behave like `pure` and `const` in C++ respectively. 
+If the distinction between `view` and `pure` is confusing for smart contract developers learning 
+Cadence with some past Solidity experience, we might consider using `view` instead of `pure` for
+this new syntax. 
+
+Swift takes the opposite approach, having a `mutating` keyword that functions must use in order to
+be allowed to mutate program state, essentially implying that functions are `pure` by default. Given
+the large amount of existing code in Cadence however, adoption will be easier by default if impurity 
+is the default assumption. 
+
+## Questions and Discussion Topics
+
+* Should `pure` functions be allowed to emit events and log data? These are technically side effects, but
+they are not problematic for the specific use cases we have identified for this analysis. Are there any other
+potential use cases for this analysis that would require events and logging to be restricted in `pure` contexts?
+
+* Given that `pure` in Cadence is closer to `view` in Solidity, a language with which our users may be familiar, 
+should we instead consider using `view` instead of `pure` in our implementation of this feature?
```

Comment:
2 cents: I quite like Solidity's keywords/definitions, as they match more closely the typical definitions of what is considered a pure function, I'd go with `view`

Reply:
cc @SupunS @robert-e-davidson3 @bluesign @bjartek how do you all feel about this. I am ambivalent either way, but I want to get buy in from a majority of the FLIP commenters before i make the switch.

Reply:
`view` seems better for me personally.